controller/hybridgateway/reconciler_utils.go

```diff
@@ -5,163 +5,178 @@ import (
+	// Get the desired state from the converter.
+	desiredObjects := conv.GetOutputStore(ctx)
+	if len(desiredObjects) == 0 {
+		logger.V(1).Info("No desired objects to enforce")
+		return false, false, nil
 	}
 
-	// Create a map of the owned resources using the hash spec as index.
-	ownedResourceMap := mapOwnedResources(rootObjectPtr, resources)
-	for _, expectedObject := range store {
-		expectedHash := expectedObject.GetLabels()[consts.GatewayOperatorHashSpecLabel]
-		existingObject, found := ownedResourceMap[expectedHash]
-		if found {
-			existingObject.hits++
-		} else {
-			if err := cl.Create(ctx, &expectedObject); err != nil {
-				if errors.IsAlreadyExists(err) {
-					continue
+	for _, desired := range desiredObjects {
+		// Get the existing object by name from the API server.
+		existing := &unstructured.Unstructured{}
+		existing.SetGroupVersionKind(desired.GetObjectKind().GroupVersionKind())
+
+		err := cl.Get(ctx, client.ObjectKey{
+			Namespace: desired.GetNamespace(),
+			Name:      desired.GetName(),
+		}, existing)
+
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Object doesn't exist, create it using server-side apply.
+				logger.V(1).Info("Creating new object", "kind", desired.GetKind(), "name", desired.GetName(), "namespace", desired.GetNamespace())
+
+				// Set field manager for server-side apply
+				if err := cl.Patch(ctx, &desired, client.Apply, client.FieldOwner(FieldManager), client.ForceOwnership); err != nil {
+					if errors.IsConflict(err) {
+						logger.V(1).Info("Conflict during create, will retry", "error", err)
+						return true, false, nil
+					}
+					return true, false, fmt.Errorf("failed to create object kind %s name %s: %w", desired.GetKind(), desired.GetName(), err)
 				}
+				continue
+			} else {
+				// Other error getting the object.
+				logger.Error(err, "Failed to get existing object", "kind", desired.GetKind(), "name", desired.GetName())
+				return true, false, err
 			}
+		}
+
+		// Handle the case when resource are marked for deletion.
+		if !existing.GetDeletionTimestamp().IsZero() {
+			logger.V(1).Info("Existing object is marked for deletion, will not enforce state", "kind", existing.GetKind(), "name", existing.GetName())
 			continue
 		}
 
-		// TODO: ensure the spec is up to date. This print is meant
-		// to act as a placeholder for the actual update logic.
-		// https://github.com/Kong/kong-operator/issues/2171
-		logger.Info("TODO: ensure the spec is up to date for", "name", expectedObject.GetName())
-	}
+		// Object exists, check if we need to update it.
+		managedFieldsObj, err := managedfields.ExtractAsUnstructured(existing, FieldManager, "")
+		if err != nil {
+			return true, false, fmt.Errorf("failed to extract managed fields for kind %s name %s: %w", existing.GetKind(), existing.GetName(), err)
+		}
+		if managedFieldsObj == nil {
+			// No managed fields for our field manager, we should update.
```

Comment:
This is the case when someone else created the child resource and we are going to overwrite and take control of it, right?